internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -717,6 +696,29 @@ func (r *BMACReconciler) findBMH(ctx context.Context, agent *aiv1beta1.Agent) (*
 	return nil, nil
 }
 
+// Find `BareMetalHost` resources that match an InfraEnv
+//
+// Only `BareMetalHost` resources that have a label with a
+// reference to an InfraEnv
+func (r *BMACReconciler) findBMHByInfraEnv(ctx context.Context, infraEnv *aiv1beta1.InfraEnv) ([]*bmh_v1alpha1.BareMetalHost, error) {
+	bmhList := bmh_v1alpha1.BareMetalHostList{}
+	err := r.Client.List(ctx, &bmhList, client.InNamespace(infraEnv.Namespace))
+	if err != nil {
+		return nil, err
+	}
+
+	bmhs := []*bmh_v1alpha1.BareMetalHost{}
+
+	for _, bmh := range bmhList.Items {
+		if val, ok := bmh.ObjectMeta.Labels[BMH_INFRA_ENV_LABEL]; ok {
+			if strings.EqualFold(val, infraEnv.Name) {
+				bmhs = append(bmhs, &bmh)
```

Comment:
I think you've fallen victim to [this gotcha](https://www.evanjones.ca/go-gotcha-loop-variables.html). The fix is to use the index value, and change this to:

```
bmhs = append(bmhs, &bmhList.Items[i])
```

Reply:
:facepalm:  thanks